File: LeetCode/Top Interview Questions/Strings/Medium/longest_palindromic_substring_hari.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        if(s.length()==1)
        {
            return s;
        }
        if(s.length()==2)
        {
            if(s[0]==s[1])
            {
                return s;
            }
            else
            {
                return s.substr(0,1);
            }
        }
        if(s.length()<1)
        {
            return "";
        }
        int start=0;
        int end=0;
        int i=0;
        for(i=0;i<s.length();i++)
        {
            int l1=middle(s,i,i);
            int l2=middle(s,i,i+1);
            int len = l1>l2? l1:l2;
            if(len>end-start)
            {
                start = i - (len - 1) / 2;
                end = i + len / 2;  
            }
            
        }
        return s.substr(start,end+1);
        
    }
    
    int middle(string s , int left , int right)
    {
        int l=left;
        int r=right;
        while(l>=0 && r<s.length() && s[l]==s[r])
        {
            l--;
            r++;
        }
        return r-l-1;
    }
};

File: LeetCode/Mock Interviews/Strings/longest_palindromic_substring_hari.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n=len(s)
        dp=[[0 for x in range(n)] for y in range(n)]
        max_len=1
        i=0
        while(i<n):
            dp[i][i]=True
            i=i+1
        start=0
        i=0
        while(i<n-1):
            if(s[i]==s[i+1]):
                dp[i][i+1]=True
                start=i
                max_len=2
            i+=1
        k=3
        while(k<=n):
            i=0
            while(i<n-k+1):
                j=i+k-1
                if(dp[i+1][j-1]==True and s[i]==s[j]):
                    dp[i][j]=True
                    if(k>max_len):
                        start=i
                        max_len=k
                i+=1
            k+=1
        end=start+max_len-1
        return s[start:end+1]
